### PR TITLE
[NETBEANS-1147] Better enablement of fragment modules.

### PR DESCRIPTION
**Pull request by sdedic**
Inspired by #715 change with additional tests provided.

Originally the branch should only add tests, but as it showed up, some corner cases were not handled (I think) still well.

I wanted to cover 3 cases: 
1. the user enables a module which is a fragment; which should imply enabling the host module as well. 2.  if the user enables the hosting module which just happens to have some fragments around, all those will be enabled and injected (if dependencies are satisfied)
3. if a fragment is enabled, but its host module is **already loaded and live**, the fragment cannot (reliably) join the constructed classloader. Exception should be thrown.

- I found a place in `AutoupdateCatalogParser` where the `needsRestart` condition is probably wrong: if the module is a fragment, then the restart is always needed. Maybe could be refined more in that the host module must be already installed & enabled.

- in original code, the fragments were registered at the start of `ModuleManager.enable`, so that `simulateEnable` would not report them. I changed that so `ModuleManager` collects fragment list immediately as fragment `Module` objects are created. Then it can always find all fragments for a module.

### CODE REVIEW

File: platform/autoupdate.services/src/org/netbeans/modules/autoupdate/updateprovider/AutoupdateCatalogParser.java [hunk 1]
```diff
@@ -545,7 +545,11 @@ public void appendModuleAttributes (Attributes module) {
             String autoload = module.getValue (MODULE_ATTR_AUTOLOAD);
             String preferred = module.getValue(MODULE_ATTR_IS_PREFERRED_UPDATE);
                         
-            needsRestart = isFragment || needsrestart == null || needsrestart.trim ().length () == 0 ? null : Boolean.valueOf (needsrestart);
+            if (isFragment) {
+                needsRestart = true;
```
Comment (JaroslavTulach) [inline platform/autoupdate.services/src/org/netbeans/modules/autoupdate/updateprovider/AutoupdateCatalogParser.java]: Will that work with #715? The core idea is that we can install NbJavac (which includes a fragment) without restart!

File: platform/o.n.bootstrap/test/unit/src/org/netbeans/ModuleManagerTest.java [hunk 2]
```diff
@@ -2760,6 +2760,57 @@ public void testMissingSpecVersion() throws Exception {
         Module client = mgr.create(jar, null, false, false, false);
         assertEquals(1, client.getProblems().size());
     }
+    
+    public void testEnableFragmentBeforeItsHost() throws Exception {
```
Comment (JaroslavTulach) [inline platform/o.n.bootstrap/test/unit/src/org/netbeans/ModuleManagerTest.java]: Nice tests.

File: platform/autoupdate.services/src/org/netbeans/modules/autoupdate/updateprovider/AutoupdateCatalogParser.java (hunk 1 above)
Comment (sdedic) [reply platform/autoupdate.services/src/org/netbeans/modules/autoupdate/updateprovider/AutoupdateCatalogParser.java]: OK, let's take this change to #715. It just improves consistency with line ~570; I believe that  I added the isFragment condition erroneously here (parenthesis missing). Maybe better condition is needed: in fact restart is only needed if the hosting module is present/enabled prior to autoupdate operation.

Comment (JaroslavTulach) [reply platform/autoupdate.services/src/org/netbeans/modules/autoupdate/updateprovider/AutoupdateCatalogParser.java]: If "restart is only needed if the hosting module has already been present", then everything is OK.

**Comment by sdedic** [comment]
rebased on latest master merged with other changes in o.n.bootstrap
